File: sdk/go/spl/crypto.go
package spl

import (
	"crypto/ed25519"
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
)

// VerifyEd25519 checks an Ed25519 signature over a message.
func VerifyEd25519(message []byte, signatureHex, publicKeyHex string) bool {
	sig, err := hex.DecodeString(signatureHex)
	if err != nil || len(sig) != ed25519.SignatureSize {
		return false
	}
	pub, err := hex.DecodeString(publicKeyHex)
	if err != nil || len(pub) != ed25519.PublicKeySize {
		return false
	}
	return ed25519.Verify(ed25519.PublicKey(pub), message, sig)
}

// SHA256Hash returns the SHA-256 hash of data.
func SHA256Hash(data []byte) []byte {
	h := sha256.Sum256(data)
	return h[:]
}

// MerkleProofStep represents one step in a Merkle proof.
type MerkleProofStep struct {
	Hash     string `json:"hash"`
	Position string `json:"position"` // "left" or "right"
}

// VerifyMerkleProof checks a Merkle proof for leafData against rootHex.
func VerifyMerkleProof(leafData string, proof []MerkleProofStep, rootHex string) bool {
	current := SHA256Hash([]byte(leafData))

	for _, step := range proof {
		sibling, err := hex.DecodeString(step.Hash)
		if err != nil {
			return false
		}
		h := sha256.New()
		if step.Position == "right" {
			h.Write(current)
			h.Write(sibling)
		} else {
			h.Write(sibling)
			h.Write(current)
		}
		current = h.Sum(nil)
	}

	return hex.EncodeToString(current) == rootHex
}

// HashTuple hashes a slice of values by JSON-serializing then SHA-256.
func HashTuple(tuple []any) string {
	b, err := json.Marshal(tuple)
	if err != nil {
		return ""
	}
	h := sha256.Sum256(b)
	return hex.EncodeToString(h[:])
}

// hkdfSHA256 implements HKDF-SHA256 (RFC 5869) extract-and-expand.
// Zero external dependencies.
func hkdfSHA256(ikm, salt, info []byte, length int) []byte {
	// Extract: PRK = HMAC-SHA256(salt, ikm)
	if len(salt) == 0 {
		salt = make([]byte, sha256.Size)
	}
	mac := hmac.New(sha256.New, salt)
	mac.Write(ikm)
	prk := mac.Sum(nil)

	// Expand: output = T(1) || T(2) || ...
	var out []byte
	var prev []byte
	for i := 1; len(out) < length; i++ {
		mac = hmac.New(sha256.New, prk)
		mac.Write(prev)
		mac.Write(info)
		mac.Write([]byte{byte(i)})
		prev = mac.Sum(nil)
		out = append(out, prev...)
	}
	return out[:length]
}

// DeriveServiceKey derives a service-specific Ed25519 keypair using HKDF-SHA256.
// Provides unlinkability: different services see different public keys.
func DeriveServiceKey(masterKeyHex, serviceDomain string) (publicKeyHex, privateKeyHex string, err error) {
	masterKey, err := hex.DecodeString(masterKeyHex)
	if err != nil {
		return "", "", err
	}
	salt := []byte("agent-safe-v1")
	info := []byte(serviceDomain)
	seed := hkdfSHA256(masterKey, salt, info, ed25519.SeedSize)

	priv := ed25519.NewKeyFromSeed(seed)
	pub := priv.Public().(ed25519.PublicKey)
	return hex.EncodeToString(pub), hex.EncodeToString(seed), nil
}

// VerifyHashChain checks that hashing preimageHex (chainLength - index) times
// produces the commitment.
func VerifyHashChain(commitment, preimageHex string, index, chainLength int) bool {
	current, err := hex.DecodeString(preimageHex)
	if err != nil {
		return false
	}
	steps := chainLength - index
	for i := 0; i < steps; i++ {
		h := sha256.Sum256(current)
		current = h[:]
	}
	return hex.EncodeToString(current) == commitment
}

File: sdk/python/spl/crypto.py
"""Real crypto implementations for SPL token verification.

Ed25519: requires `cryptography` package (optional dependency).
Merkle/hash-chain: uses stdlib hashlib only.
"""

import hashlib
import json
from typing import Any


def sha256(data: bytes) -> bytes:
    """SHA-256 hash of data."""
    return hashlib.sha256(data).digest()


def sha256_hex(data: bytes) -> str:
    """SHA-256 hash of data as hex string."""
    return hashlib.sha256(data).hexdigest()


def verify_ed25519(message: bytes, signature_hex: str, public_key_hex: str) -> bool:
    """Verify an Ed25519 signature over a message.

    Requires the `cryptography` package.
    """
    try:
        from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PublicKey
        pub_bytes = bytes.fromhex(public_key_hex)
        sig_bytes = bytes.fromhex(signature_hex)
        public_key = Ed25519PublicKey.from_public_bytes(pub_bytes)
        public_key.verify(sig_bytes, message)
        return True
    except ImportError:
        raise RuntimeError(
            "Ed25519 verification requires the 'cryptography' package. "
            "Install with: pip install agent-safe-spl[crypto]"
        )
    except Exception:
        return False


def verify_merkle_proof(
    leaf_data: str,
    proof: list[dict[str, str]],
    root_hex: str,
) -> bool:
    """Verify a Merkle proof for leaf_data against root_hex.

    Args:
        leaf_data: The leaf value (will be SHA-256 hashed)
        proof: List of {"hash": hex, "position": "left"|"right"}
        root_hex: Expected root hash in hex
    """
    try:
        current = sha256(leaf_data.encode("utf-8"))

        for step in proof:
            sibling = bytes.fromhex(step["hash"])
            if step["position"] == "right":
                current = sha256(current + sibling)
            else:
                current = sha256(sibling + current)

        return current.hex() == root_hex
    except Exception:
        return False


def hash_tuple(values: list[Any]) -> str:
    """Hash a tuple by JSON-serializing then SHA-256."""
    serialized = json.dumps(values, separators=(",", ":"))
    return sha256_hex(serialized.encode("utf-8"))


def verify_hash_chain(
    commitment: str,
    preimage_hex: str,
    index: int,
    chain_length: int,
) -> bool:
    """Verify a hash chain receipt.

    Hashes preimage_hex (chain_length - index) times and checks
    it equals the commitment.
    """
    try:
        current = bytes.fromhex(preimage_hex)
        steps = chain_length - index

        for _ in range(steps):
            current = sha256(current)

        return current.hex() == commitment
    except Exception:
        return False
